metaflow/datatools/s3op.py

```diff
@@ -106,66 +157,125 @@ def worker(queue, mode):
+                do_upload = False
+                if pre_op_info:
+                    result_info = op_info(url)
+                    if result_info['error'] == ERROR_URL_NOT_FOUND:
+                        # We only upload if the file is not found
+                        do_upload = True
+                else:
+                    # No pre-op so we upload
+                    do_upload = True
+                if do_upload:
+                    extra=None
+                    if url.content_type or url.metadata:
+                        extra = {}
+                        if url.content_type:
+                            extra['ContentType'] = url.content_type
+                        if url.metadata is not None:
+                            extra['Metadata'] = url.metadata
+                    s3.upload_file(url.local, url.bucket, url.path, ExtraArgs=extra)
+                    # We indicate that the file was uploaded
+                    result_pipe.send((idx, 0))
 
     except:
         traceback.print_exc()
         sys.exit(ERROR_WORKER_EXCEPTION)
 
 def start_workers(mode, urls, num_workers):
-
+    # We start the minimum of len(urls) or num_workers to avoid starting
+    # workers that will definitely do nothing
+    num_workers = num_workers if len(urls) > num_workers else len(urls)
     queue = Queue(len(urls) + num_workers)
     procs = {}
 
     # 1. push sources and destinations to the queue
-    for url, _ in urls:
-        queue.put(url)
+    idx = 0
+    for elt in urls:
+        queue.put((elt, idx))
+        idx += 1
 
     # 2. push end-of-queue markers
     for i in range(num_workers):
-        queue.put(None)
+        queue.put((None, None))
+
+    # 3. Prepare the result structure
+    sz_results = [None]*idx
 
-    # 3. start processes
+    # 4. start processes
     for i in range(num_workers):
-        p = Process(target=worker, args=(queue, mode))
+        recv, send = Pipe(duplex=False)
+        p = Process(target=worker, args=(send, queue, mode))
         p.start()
-        procs[p] = True
-
-    # 4. wait for the processes to finish
-    while any(procs.values()):
-        for proc, is_alive in procs.items():
-            if is_alive:
-                proc.join(timeout=1)
-                if proc.exitcode is not None:
-                    if proc.exitcode == 0:
-                        procs[proc] = False
-                    else:
-                        msg = 'Worker process failed (exit code %d)'\
-                               % proc.exitcode
-                        exit(msg, proc.exitcode)
+        procs[p] = recv
+        # Do not keep references to the Connection around to make sure
+        # they close properly
+        recv = None
+        send = None
+
+    # 5. wait for the processes to finish; we continuously update procs
+    # to remove all processes that have finished already
+    while procs:
+        new_procs = {}
+        for proc, recv in procs.items():
+            proc.join(timeout=1)
```

Comment:
this call will block for a second while workers are executing. If means that if you have, say, 32 workers (the default), it will take 31 seconds before we get to the last worker. It is possible that a worker will have filled its pipe buffer during this long wait time and it starts blocking, which hurts performance.

Reply:
See above but yes, it would wait. I could set a quicker timeout and have this be more of a busy loop.

Reply:
A downside of heuristics like this (1 second or less?) and the 10 below is that they can cause performance degradations that are nightmarish to identify and debug.

The "correct" way of handling the read side would be to make pipes non-blocking and use e.g. epoll to multiplex input from pipes. It gets tedious. Hence my suggestion above to consider if there's a way to get rid of pipes altogether.

Reply:
See above for the elimination of pipes. For the 1 second poll, the timeout was also present in the initial code. I am perfectly fine removing it and putting a tighter loop but it will always be a heuristical approach. Another way we could do it is have the process signal that it is done (but we are back to multiplexing) but that seems a bit overkill.

Reply:
Kept 1 second poll but pipe removed.